assets/shared/helpers/player.js

```diff
@@ -0,0 +1,552 @@
+					! event.data.currentTimeMs
+				) {
+					return;
+				}
+				callback( event.data.currentTimeMs / 1000 );
+			};
+
+			// eslint-disable-next-line @wordpress/no-global-event-listener -- Not in a React context.
+			window.addEventListener( 'message', transformedCallback );
+
+			return () => {
+				// eslint-disable-next-line @wordpress/no-global-event-listener -- Not in a React context.
+				window.removeEventListener( 'message', transformedCallback );
+			};
+		},
+	},
+	[ YOUTUBE_TYPE ]: {
+		/**
+		 * The embed pattern to check if it's the respective type.
+		 */
+		embedPattern: /(youtu\.be|youtube\.com)\/.+/i,
+
+		/**
+		 * Initialize the player.
+		 *
+		 * @param {HTMLIFrameElement} element The player element.
+		 *
+		 * @return {Object} The YouTube player instance through a promise.
+		 */
+		initializePlayer: ( element ) =>
+			new Promise( ( resolve ) => {
+				window.senseiYouTubeIframeAPIReady.then( () => {
+					const player =
+						window.YT.get( element.id ) ||
+						new window.YT.Player( element );
+
+					const onReady = () => {
+						resolve( player );
+					};
+
+					if ( player.getDuration ) {
+						// Just in case it's called after the player is ready.
+						onReady();
+					} else {
+						player.addEventListener( 'onReady', onReady );
+					}
+				} );
+			} ),
+
+		/**
+		 * Get the video duration.
+		 *
+		 * @param {Object} player The YouTube player instance.
+		 *
+		 * @return {Promise<number>} The duration of the video in seconds through a promise.
+		 */
+		getDuration: ( player ) => Promise.resolve( player.getDuration() ),
+
+		/**
+		 * Set the video to a current time.
+		 *
+		 * @param {Object} player  The YouTube player instance.
+		 * @param {number} seconds The video time in seconds to set.
+		 *
+		 * @return {Promise} A resolved promise.
+		 */
+		setCurrentTime: ( player, seconds ) => {
+			player.seekTo( seconds );
+			return Promise.resolve();
+		},
+
+		/**
+		 * Play the video.
+		 *
+		 * @param {Object} player The YouTube player instance.
+		 *
+		 * @return {Promise} A resolved promise.
+		 */
+		play: ( player ) => {
+			player.playVideo();
+			return Promise.resolve();
+		},
+
+		/**
+		 * Pause the video.
+		 *
+		 * @param {Object} player The YouTube player instance.
+		 *
+		 * @return {Promise} A resolved promise.
+		 */
+		pause: ( player ) => {
+			player.pauseVideo();
+			return Promise.resolve();
+		},
+
+		/**
+		 * Add an event listener to the player.
+		 *
+		 * @param {Object}   player    The YouTube player instance.
+		 * @param {string}   eventName Event name (currently only `timeupdate` is supported)
+		 * @param {Function} callback  Listener callback.
+		 *
+		 * @return {Function} The function to unsubscribe the event.
+		 */
+		on: ( player, eventName, callback ) => {
+			const timer = 250;
+
+			const interval = setInterval( () => {
+				if (
+					player.getPlayerState() === window.YT.PlayerState.PLAYING
+				) {
+					callback( player.getCurrentTime() );
```

Comment:
Do you think it would make sense to remember the last currentTime and fire only if the time changed?

I am not too concerned but with a 250ms interval we might be calling the callback up to 4 times per second with the same time.

Reply:
Let me know if I missed something, but it shouldn't be the same time. It should return the float number for the second, like this:

<img width="268" alt="Screen Shot 2022-07-20 at 15 39 07" src="https://user-images.githubusercontent.com/876340/180058007-e0a9fa3f-1ed0-430e-954f-3131e12a9050.png">

Reply:
What about when the video is paused? Wouldn't this continue calling the callback with the same time in that case?

Reply:
> What about when the video is paused? Wouldn't this continue calling the callback with the same time in that case?

It would continue calling the `setIntervall` callback, but not our `callback` because of the `if`. We could make some different implementation to start the interval and stop watching the video state, but it would be a little more complex, and make the returned unsubscribe much more complex too. WDYT?

IMO, since it wouldn't have a big impact on performance, I'd continue with this as it is to avoid complex code.

Reply:
Just for clarification: when I started this thread I thought currentTime was an integer in seconds. So that's why I was concerned that it would be called with the same integer multiple times.
Considering that this is a float I am ok with this originally :)